### PR TITLE
sql: add approx_disk_bytes to SHOW RANGES WITH DETAILS

### LINKED ISSUE #97858: sql,kv: extend SHOW RANGES to include table/index size estimates
**Describe the problem**

Currently `SHOW RANGES WITH DETAILS` only includes range sizes, not object sizes.

When a range contains multiple table/indexes, the range size cannot be used to estimate the table/index size any more.

So we need a replacement for where people would previously use SUM on the range size.

This can use the new span stats computation that Zach implemented in #96223 .

Jira issue: CRDB-24925
Epic: CRDB-24928

### PR DESCRIPTION
**Pull request by Santamaura**
This change adds an approx_disk_bytes column to SHOW RANGES WITH DETAILS.
The column is derived from the approximate_disk_bytes column from
crdb_internal.tenant_span_stats generator builtin.
The generator function needed to be updated in order for this to work
therefore, it has been updated to be capable of accepting db name and
table name as parameters.
The approx_disk_bytes metric  can be used to estimate the total size
of the tables within the scope of the query.

Fixes: https://github.com/cockroachdb/cockroach/issues/97858

Release note (sql change): add approx_disk_bytes column to
SHOW RANGES WITH DETAILS and update crdb_internal.tenant_span_stats
to accept db name and table name strings.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Santamaura)

___
*[`pkg/sql/delegate/show_ranges.go` line 378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQBrRk89H9OafHM_gL0:-NQBrRk89H9OafHM_gL1:b-cepwxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/05ef50eb0fcd48aa56adcba2374547006997fc39/pkg/sql/delegate/show_ranges.go#L378)):*
> ```Go
> 		buf.WriteString("span_stats AS (SELECT sum(approximate_disk_bytes) AS approx_disk_bytes FROM crdb_internal.tenant_span_stats(")
> 		if dbName != "" {
> 			stmt := fmt.Sprintf(`SELECT parent_id FROM crdb_internal.tables WHERE database_name = '%s' LIMIT 1`, dbName.String())
> ```

We will need this to be more clever and expand tenant_span_stats as part of the surrounding query (no recursive call to QueryRowEx).

You can use `SHOW RANGES WITH EXPLAIN` to familiarize yourself with the shape of the query and find the right location for the call.

File: pkg/sql/delegate/show_ranges.go [hunk 1]
```diff
@@ -367,13 +369,29 @@ AND s.end_key > r.start_key`)
               ON r.table_id = t.table_id`, dbName.String(), dbNameCol)
 	}
 	buf.WriteString("\n),\n") // end of named_ranges CTE.
+	// If with details we need to get approximate_disk_bytes from tenant_span_stats.
+	if n.Options.Details {
+		buf.WriteString("span_stats AS (SELECT sum(approximate_disk_bytes) AS approx_disk_bytes FROM crdb_internal.tenant_span_stats(")
```
Comment (knz) [inline pkg/sql/delegate/show_ranges.go:374]: Are you sure this is what we want? What if a range contains data from multiple tables? If I run `SHOW RANGE WITH TABLES` on a range containing multiple tables, i probably want to see the approx size for each of them.

Comment (Santamaura) [reply pkg/sql/delegate/show_ranges.go:374]: Ah thanks for pointing this out, I guess in that case we would not want to sum the rows.

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @koorosh, @Santamaura, and @zachlite)

___
*[`pkg/sql/planner.go` line 861 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQ_vRiWEMe_JOnnqHPD:-NQ_vRiX7ArGioeUFDWR:b41bfwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c6fdb270f60ba0c497bfc671448ea78da47528/pkg/sql/planner.go#L861)):*
> ```Go
> // database/table.
> func (p *planner) GetDetailsForSpanStats(
> 	ctx context.Context, dbId int, tableId int, getNamesForQuery func() (string, string),
> ```

If `getNamesForQuery` returns `dbName`, `tableName` with no input can you make those into `GetDetailsForSpanStats` params?

**Review by Santamaura — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @knz, @koorosh, and @zachlite)

___
*[`pkg/sql/planner.go` line 861 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQ_vRiWEMe_JOnnqHPD:-NQaLb-NE90XHdnMhyQy:bxmgcxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c6fdb270f60ba0c497bfc671448ea78da47528/pkg/sql/planner.go#L861)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

If `getNamesForQuery` returns `dbName`, `tableName` with no input can you make those into `GetDetailsForSpanStats` params?
</blockquote></details>

If there is no input we would just do a query where `dbId` is non zero (assuming no `dbId`/`tableId` either). I'm not sure if this answers your question, if it doesn't can you clarify what you're asking?

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @koorosh, @Santamaura, and @zachlite)

___
*[`pkg/sql/planner.go` line 861 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQ_vRiWEMe_JOnnqHPD:-NQaf1F7E5VwIY5kRXom:b-ipzpls) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c6fdb270f60ba0c497bfc671448ea78da47528/pkg/sql/planner.go#L861)):*
<details><summary><i>Previously, Santamaura (Alex Santamaura) wrote…</i></summary><blockquote>

If there is no input we would just do a query where `dbId` is non zero (assuming no `dbId`/`tableId` either). I'm not sure if this answers your question, if it doesn't can you clarify what you're asking?
</blockquote></details>

Instead of using a `func() (string, string)`, you can change the function signature to
```
func (p *planner) GetDetailsForSpanStats(
	ctx context.Context, dbId int, tableId int, dbName string, tableName string,
) (eval.InternalRows, error)
```
so that the `*name` params are consistent with the `*id` params.

**Review by Santamaura — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @knz, @koorosh, and @zachlite)

___
*[`pkg/sql/planner.go` line 861 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQ_vRiWEMe_JOnnqHPD:-NQb-soP76KTBMZu1HWZ:btoedfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c6fdb270f60ba0c497bfc671448ea78da47528/pkg/sql/planner.go#L861)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Instead of using a `func() (string, string)`, you can change the function signature to
```
func (p *planner) GetDetailsForSpanStats(
	ctx context.Context, dbId int, tableId int, dbName string, tableName string,
) (eval.InternalRows, error)
```
so that the `*name` params are consistent with the `*id` params.




</blockquote></details>

Ah ok, done.

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @koorosh, @Santamaura, and @zachlite)

___
*[`pkg/sql/delegate/show_ranges.go` line 350 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQeopihE-hEkHvAbu62:-NQeopihE-hEkHvAbu63:b7amasx) ([raw file](https://github.com/cockroachdb/cockroach/blob/283fc2dbcbaf662a83affccd8e8cb70b36d140f6/pkg/sql/delegate/show_ranges.go#L350)):*
> ```Go
> 		if dbName != "" {
> 			if isFromTableOrIndex {
> 				tenantSpanStats += fmt.Sprintf(`'%s', '%s'`, dbName, string(idx.Table().Name()))
> ```

So what if I run `SHOW RANGES WITH INDEXES`? Will it give me the cumulative size for the entire table on each index row? Don't we want the approx size of each index on different rows instead?
___
*[`pkg/sql/delegate/show_ranges.go` line 368 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQeoyc581ajKB_or6p_:-NQeoyc6AtqYpXtM4eIE:bzeznf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/283fc2dbcbaf662a83affccd8e8cb70b36d140f6/pkg/sql/delegate/show_ranges.go#L368)):*
> ```Go
> 	case tree.ExpandTables:
> 		// Add table name details. Also add approximate_disk_bytes if WITH DETAILS is specified.
> 		expr := fmt.Sprintf("SELECT r.* %s, t.schema_name, t.name AS table_name ", dbNameCol)
> ```

You can use `buf.WriteString` (or `fmt.Fprintf(&buf, ...)`) throughout - no need to allocate separate strings here.

**Review by Santamaura — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @knz, @koorosh, @THardy98, and @zachlite)

___
*[`pkg/sql/delegate/show_ranges.go` line 350 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQeopihE-hEkHvAbu62:-NQfm_VZ5tdnoTr28WV_:bp52cuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/283fc2dbcbaf662a83affccd8e8cb70b36d140f6/pkg/sql/delegate/show_ranges.go#L350)):*
> So what if I run SHOW RANGES WITH INDEXES? Will it give me the cumulative size for the entire table on each index row? 

Currently yes.

> Don't we want the approx size of each index on different rows instead?

We do. This would require the `tenant_span_stats` to have a parameter to indicate that we want to call `SpanStats` with an index id instead of table id. I talked async with @THardy98 and it sounds like this PR will include those changes https://github.com/cockroachdb/cockroach/pull/98490
___
*[`pkg/sql/delegate/show_ranges.go` line 368 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98339#-NQeoyc581ajKB_or6p_:-NQfneSW-s6H7xo82CS5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/283fc2dbcbaf662a83affccd8e8cb70b36d140f6/pkg/sql/delegate/show_ranges.go#L368)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

You can use `buf.WriteString` (or `fmt.Fprintf(&buf, ...)`) throughout - no need to allocate separate strings here.
</blockquote></details>

Done.

**Pull request closed without merging**